src/hotspot/share/runtime/mrswMutex.hpp

```diff
@@ -0,0 +1,171 @@
+// current readers to proceed. The consequences are undefined if one or more
+// writer threads attempt to enter their critical region when another writer
+// thread already is attempting to do so or currently is in its critical region.
+class MRSWMutex : public CHeapObj<mtSynchronizer> {
+private:
+  NONCOPYABLE(MRSWMutex);
+  class NoTransition : public StackObj {
+  public:
+    NoTransition(Thread* thread) {
+    }
+  };
+
+  template<typename T>
+  class Locker : public StackObj {
+  private:
+    T* const _lock;
+
+  public:
+    Locker(T* lock)
+      : _lock(lock) {
+      if (_lock != nullptr) {
+        _lock->lock();
+      }
+    }
+    ~Locker() {
+      if (_lock != nullptr) {
+        _lock->unlock();
+      }
+    }
+  };
+
+  PlatformMonitor _mon;
+
+  // The count reflects the number of reader threads inside a critical region and whether or not a writer is waiting.
+  //
+  // * Normal (count >= 0). Readers are allowed to enter and exit their critical region, no writer waiting.
+  //
+  // * Blocked (count == -1). A writer is inside its critical region.
+  //
+  // * Block in progress (count < -1). Readers are only allowed to exit their critical region.
+  //   Attempts by readers to enter their critical region is blocked.
+  //
+  volatile int64_t _count;
+
+public:
+  MRSWMutex()
+    : _mon(),
+      _count(0) {
+  }
+
+  ~MRSWMutex() {
+  }
+
+  template<typename WriterTransition = NoTransition>
+  void write_lock() {
+    for (;;) {
+      const int64_t count = Atomic::load_acquire(&_count);
+
+      if (count < 0) {
+        // Already blocked, wait until unblocked
+
+        // Do requested transition before blocking
+        WriterTransition tbivm(JavaThread::current());
+
+        Locker<PlatformMonitor> locker(&_mon);
+        while (Atomic::load_acquire(&_count) < 0) {
+          _mon.wait(0);
+        }
+
+        // Unblocked
+        continue;
+      }
+
+      // Increment and invert count
+      if (Atomic::cmpxchg(&_count, count, -(count + 1)) != count) {
+        continue;
+      }
+
+      // If the previous count was 0, then we just incremented and inverted
+      // it to -1 and have now blocked. Otherwise we wait until all reader
+      // threads have exited the critical region.
+      if (count != 0) {
+        // Do requested transition before blocking
+        WriterTransition tbivm(JavaThread::current());
+
+        // Wait until blocked
+        Locker<PlatformMonitor> locker(&_mon);
+        while (Atomic::load_acquire(&_count) != -1) {
+          _mon.wait(0);
+        }
+      }
+
+      // Blocked.
+      return;
+    }
+  }
+  void write_unlock();
+
+  template<typename ReaderTransition = NoTransition>
+  void read_lock() {
+    for (;;) {
+      const int64_t count = Atomic::load_acquire(&_count);
+      if (count < 0) {
+        // Wait until unblocked
+
+        // Do requested transition before blocking
+        ReaderTransition tbivm(JavaThread::current());
```

Comment:
This shouldn't compute the thread, instead `read_lock` and `write_lock` should be given the threads as arguments. This will fail if the current thread isn't a Java one.